### PR TITLE
Support for ROI Pooling

### PR DESCRIPTION
**Pull request by varunagrawal**
ROI Pooling with tests. Fix for cuda context in ROI Align.

@fmassa please can you review this?

### CODE REVIEW

File: torchvision/csrc/cuda/ROIPool_cuda.cu [hunk 1]
```diff
@@ -193,7 +197,11 @@ at::Tensor ROIPool_backward_cuda(const at::Tensor& grad,
          pooled_height,
          pooled_width,
          grad_input.data<scalar_t>(),
-         rois.data<scalar_t>());
+         rois.data<scalar_t>(),
+         n_stride,
```
Comment (fmassa) [inline torchvision/csrc/cuda/ROIPool_cuda.cu]: interesting, so this adds support for non-contiguous gradients as well.
Did you have cases where that was the case?
It makes me think that I should probably add a `.contiguous()` call to `ROIAlign` as well

File: torchvision/csrc/cuda/ROIPool_cuda.cu [hunk 2]
```diff
@@ -75,34 +73,35 @@ __global__ void RoIPoolFForward(const int nthreads, const T* bottom_data,
...
-    T* offset_bottom_diff = bottom_diff + bottom_offset;
-    const int* offset_argmax_data = argmax_data + top_offset;
-
-    int argmax = offset_argmax_data[ph * pooled_width + pw];
-    if (argmax != -1) {
-      atomicAdd(
-          offset_bottom_diff + argmax,
-          static_cast<T>(offset_top_diff[ph * pooled_width + pw]));
-
+    const int pooled_height, const int pooled_width, T* bottom_data,
```
Comment (fmassa) [inline torchvision/csrc/cuda/ROIPool_cuda.cu]: I find this `bottom_data` name misleading, because it is actually the gradient of the `bottom`, so I think that `bottom_diff` was a better name, and more closely aligns with what was implemented in C2

File: torchvision/csrc/cuda/ROIPool_cuda.cu [hunk 3]
```diff
@@ -30,7 +28,7 @@ __global__ void RoIPoolFForward(const int nthreads, const T* bottom_data,
     int roi_end_w = round(offset_bottom_rois[3] * spatial_scale);
     int roi_end_h = round(offset_bottom_rois[4] * spatial_scale);
 
-    // Force malformed ROIs to be 1x1
+    // Force malformed ROIs to be 1x1 or HxW
```
Comment (fmassa) [inline torchvision/csrc/cuda/ROIPool_cuda.cu]: by malformed ROIs we meant ROIs with size 0 or smaller, and they become 1x1, not HxW, can you revert this change?

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 1 above)
Comment (varunagrawal) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: I encountered it when I was writing the tests since as a quick hack, I was just doing 
```
output = roi_pool(input, rois)
s = output.sum()
s.backward()
```
and this was giving me issues because the gradient was a Tensor with `numel = 1` and all strides set to 0.

The current approach I implemented was more generic and I am happy to add this to ROIAlign as well once ROIPooling is merged.

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 2 above)
Comment (varunagrawal) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: I actually wanted to update the names to be more torch-like e.g. `input`, `output`, `grad_input` and `grad_output`, since that would lead to minimum confusion. Should I just do that?

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 3 above)
Comment (varunagrawal) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: Gotcha. Updated.

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 1 above)
Comment (fmassa) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: I see.
Yes, having the strides handling is definitely more general, but as a first step I'd just add the `.contiguous()` call, and later on add support for strides.

File: torchvision/csrc/cuda/ROIPool_cuda.cu [hunk 4]
```diff
@@ -58,51 +56,51 @@ __global__ void RoIPoolFForward(const int nthreads, const T* bottom_data,
...
-      atomicAdd(
-          offset_bottom_diff + argmax,
-          static_cast<T>(offset_top_diff[ph * pooled_width + pw]));
-
+    const int pooled_height, const int pooled_width, 
+    T* grad_input, const T* rois, 
+    const int n_stride, const int c_stride,
+    const int h_stride, const int w_stride) {
+
+    CUDA_1D_KERNEL_LOOP(index, nthreads) {
```
Comment (fmassa) [inline torchvision/csrc/cuda/ROIPool_cuda.cu]: nit: for next time, try not to modify the indentation of the code, as it makes reviewing easier. Also, we are using 2 spaces for C++ files, so the previous indentation was ok. I'm merging this in anyway, but just wanted to let you know

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 1 above)
Comment (varunagrawal) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: Just to be sure, I should be calling `contiguous()` on the `grad` variable in the backwards pass at the python level, correct? This is specifically for ROI Pooling.

File: torchvision/csrc/cuda/ROIPool_cuda.cu (hunk 4 above)
Comment (varunagrawal) [reply torchvision/csrc/cuda/ROIPool_cuda.cu]: Understood, will keep that in mind.

**Comment by andfoy** [comment]
@fmassa @varunagrawal I'm unable to compile the latest revision of ``layers`` branch with the most recent PyTorch nightly wheels, it fails with the following error traceback:

```
in file included from /tmp/pip-req-build-2943ca2h/torchvision/csrc/cpu/vision.h:2:0,
                     from /tmp/pip-req-build-2943ca2h/torchvision/csrc/cpu/ROIAlign_cpu.cpp:1:
    /home/eamargffoy/anaconda3/envs/pytorch_041/lib/python3.6/site-packages/torch/lib/include/torch/csrc/api/include/torch/torch.h:13:2: warning: #warning "Including torch/torch.h for C++ extensions is deprecated. Please include torch/extension.h" [-Wcpp]
     #warning "Including torch/torch.h for C++ extensions is deprecated. Please include torch/extension.h"
      ^~~~~~~
    /tmp/pip-req-build-2943ca2h/torchvision/csrc/cpu/ROIAlign_cpu.cpp: In function ‘at::Tensor ROIAlign_forward_cpu(const at::Tensor&, const at::Tensor&, float, int, int, int)’:
    /tmp/pip-req-build-2943ca2h/torchvision/csrc/cpu/ROIAlign_cpu.cpp:234:36: error: ‘struct at::Type’ has no member named ‘tensor’; did you mean ‘renorm’?
       at::Tensor output = input.type().tensor({num_rois, channels, pooled_height, pooled_width});
                                        ^~~~~~
    error: command 'gcc' failed with exit status 1
error
Cleaning up...
``` 
Could you please give me some directions about which changes must be done in order to compile this branch without any errors?

Thanks

**Comment by varunagrawal** [comment]
@andfoy thanks for reporting this. This is an issue in ROIAlign and not ROIPool. This PR is specific to ROIPool. I am still working on fixing this. Thank you for your patience.

**Comment by andfoy** [comment]
@varunagrawal Thanks for your reply, is this error related to: https://github.com/pytorch/pytorch/pull/12360?

**Comment by varunagrawal** [comment]
Yep. That is exactly it.

**Comment by andfoy** [comment]
@varunagrawal Do you think that if all tensor creation operations are replaced by the new API, RoIAlign/Pool would work?

**Comment by fmassa** [comment]
Thanks for updating the layers branch @varunagrawal 
FYI, we will be releasing in Oct 24th an official set of layers for detection.

**Comment by fmassa** [comment]
@varunagrawal my comment was not very clear, sorry. I meant we will be releasing an object detection library, which will contain the ROI layers.

The layers will initially be part of this library. They might be moved out into torchvision at a later moment, but not initially

**Comment by varunagrawal** [comment]
Ah gotcha. I did talk to @soumith about this in 2017, so feel free to rope me in.
